app/javascript/App/bundles/common/contexts/CurrencyContext.tsx

```diff
@@ -0,0 +1,80 @@
+import * as React from 'react'
+import { withCookies, Cookies } from 'react-cookie'
+
+export interface CurrencyContextType {
+  currency: string
+  changeCurrency: (currency: string) => void
+}
+
+const CurrencyContext = React.createContext<CurrencyContextType>(null)
+
+export interface Props {
+  cookies: Cookies
+}
+
+export interface State {
+  currency: string
+}
+
+class CurrencyProvider extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props)
+
+    const { cookies } = props
+    const currency = cookies.get('currency') || 'USD'
+    this.state = {
+      currency,
+    }
+  }
+
+  public componentDidMount() {
+    document.addEventListener('currencyChange', this.onCurrencyChange)
```

Comment:
Maybe safer to make this `'currencyChange'` string a constant so the dispatcher and listener are always in sync